File: src/main/kotlin/com/edoctor/api/utils/MutexFactory.java
package com.edoctor.api.utils;

import org.hibernate.internal.util.collections.ConcurrentReferenceHashMap;
import org.springframework.stereotype.Component;

@Component
public class MutexFactory<K> {

    private ConcurrentReferenceHashMap<K, Object> map;

    public MutexFactory() {
        this.map = new ConcurrentReferenceHashMap<>();
    }

    public Object getMutex(K key) {
        return this.map.compute(key, (k, v) -> v == null ? new Object() : v);
    }

}

File: src/main/kotlin/com/edoctor/api/utils/MutexFactory.kt
package com.edoctor.api.utils

import org.hibernate.internal.util.collections.ConcurrentReferenceHashMap

class MutexFactory<K> {

    private val map: ConcurrentReferenceHashMap<K, Any> = ConcurrentReferenceHashMap()

    fun getMutex(key: K): Any = map.compute(key) { k, v -> v ?: Any() } as Any

}